battery: Initial dedicated API

aaronemassey opened the pull request:
Recreation of https://github.com/zephyrproject-rtos/zephyr/pull/44846 due to deletion of Fork. See https://github.com/zephyrproject-rtos/zephyr/pull/44846#issuecomment-1163672797

*Edit:

RFC: https://github.com/zephyrproject-rtos/zephyr/issues/44847

aaronemassey committed:
fuel_gauge: Initial fuel-gauge dedicated API

Add initial battery fuel-gauge driver API with the most basic of
native_posix driver tests.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

bjarki-trackunit commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,200 @@
+ * @param capacity	Destination for remaining time in seconds.
+ * @return non-zero if error.
+ */
+typedef int (*battery_time_to_empty)(const struct device *dev, int *seconds);
```
This function can simply return the time to empty instead of using an out param, since this time should not be able to be negative.

bjarki-trackunit commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,200 @@
+ * @return non-zero if error.
+ */
+typedef int (*battery_time_to_full)(const struct device *dev, int *seconds);
+
```
This function can return the time to empty instead of using an out param, since this time should not be able to be negative.

bjarki-trackunit commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,200 @@
+		int temperature;
+		/** BATTERY_VOLTAGE, */
+		int voltage;
+	};
```
This structs fields need better descriptions, what unit are they in?

bjarki-trackunit commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,200 @@
+ *
+ * @return non-zero if error.
+ */
+typedef int (*battery_wait_for_stable)(const struct device *dev);
```
This function needs a timeout param, otherwise the thread could block forever if there is an issue with the battery.
```
typedef int (*battery_wait_for_stable)(const struct device *dev, k_timeout timeout);
```

bjarki-trackunit commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,200 @@
+ * @return non-zero if error.
+ */
+typedef int (*battery_time_at_rate)(const struct device *dev, int rate, int *seconds);
+
```
If this function simply calculates time to empty from the battery capacity, then this function should not be an API call, it should be an inline helper, and simply return the time in seconds, without the out param.
```
static inline int (*battery_calculate_time_to_empty)(int capacity, int rate)
{
    ...
    ...
    return time_remaining;
}
```

bjarki-trackunit commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,200 @@
+ *
+ * @return non-zero if the battery driver doesn't support cut off.
+ */
+typedef int (*battery_cut_off)(const struct device *dev);
```
Remember to append _t tp typedefs battery_cut_off**_t**

bjarki-trackunit commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,200 @@
+ *
+ * @return	0 if successful, non-zero if error occurred
+ */
+typedef int (*battery_sleep_fuel_gauge)(const struct device *dev);
```
Sleep commands like this should be handled by the driver through the pm_device actions, like on/off/resume/suspend, preferably using the pm_device_runtime subsystem.

bjarki-trackunit commented on drivers/battery/Kconfig:
```diff
@@ -0,0 +1,18 @@
+module-str = battery
+source "subsys/logging/Kconfig.template.log_config"
+
+source "drivers/battery/sbs_gauge/Kconfig"
```
This source does not exist

aaronemassey replied:
? It's in this PR.

aaronemassey commented:
Just fyi, folks. I'm including tests and other work with this pull-request. I have this set as a draft for now for that reason. There's more to come!

aaronemassey commented:
> hi @aaronemassey  will you be addressing comments and feedback raised in the original PR here? :)

Absolutely. I just had to recreate the pull-request as I wanted folks to know I'm still actively working on this.

rriveramcrus commented:
Hello @aaronemassey,

This is a pretty exciting addition to Zephyr.

Some basic additions to the proposed framework to support charging ICs could be:

```
/*
 * Charging power converters can operate in the ‘charging’ mode (charging
 * and/or supplying the system), idle (converter idle), or ‘output’ mode (sending
 * power from the battery to a connected accessory).
 */
enum battery_opmode {
	BATTERY_OPMODE_IDLE = 0,
	BATTERY_OPMODE_CHARGING,
	BATTERY_OPMODE_OUTPUT,
};
```

```
/*
 * Charging systems are supplied by attached sources and can supply
 * attached sinks.
 */
enum battery_extconn_attached {
	BATTERY_NO_ATTACHMENT = 0,
	BATTERY_FIXED_SOURCE,
	BATTERY_PROG_SOURCE,
	BATTERY_FIXED_SINK,
};
```

```
enum battery_property_type {
	…
	/** Desired CC charging target in mA, 0 disables charging */
	BATTERY_DESIRED_CHARGING_CURRENT,
	/** Desired CV charging target in mV */
	BATTERY_DESIRED_CHARGING_VOLTAGE,
	/** Rising input current regulation target in mA, 0 disables regulation, -1 opens INFET */
	BATTERY_INPUT_CURRENT_REGULATION,
	/** Falling input voltage regulation target in mV, 0 disables regulation */
	BATTERY_INPUT_VOLTAGE_REGULATION
	/** Reverse mode output current limit */
	BATTERY_OUTPUT_SUPPLY_CURRENT,
	/** Reverse mode output voltage limit */
	BATTERY_OUTPUT_SUPPLY_VOLTAGE,
	/** Device operating mode */
	BATTERY_OPMODE,
	/** External connection status */
	BATTERY_EXTCONN_ATTACHED,
	…
};
```

```
struct battery_get_property {
	…
	union {
		…
		/** BATTERY_DESIRED_CHARGING_CURRENT */
		int desired_charging_current;
		/** BATTERY_DESIRED_CHARGING_VOLTAGE */
		int desired_charging_voltage;
		/** BATTERY_INPUT_CURRENT_REGULATION */
		int input_current_regulation;
		/** BATTERY_INPUT_VOLTAGE_REGULATION */
		int input_voltage_regulation;
		/** BATTERY_OUTPUT_SUPPLY_CURRENT */
		int output_supply_current;
		/** BATTERY_OUTPUT_SUPPLY_VOLTAGE */
		int output_supply_voltage;
		/** BATTERY_OPMODE */
		enum battery_opmode op_mode;
		/** BATTERY_EXTCONN_ATTACHED */
		enum battery_extconn_attached extconn_attached;
	};
};
```
Additionally charging ICs require changing settings at runtime, so a corresponding battery_set_property_t type would need to be added.

Finally, it may be a good idea to support private properties like the sensor subsystem does with the ’SENSOR_CHAN_PRIV_START’ member, but this is a ’nice-to-have’ feature.

Let me know if you have any thoughts or questions on this!
Ricardo

aaronemassey commented:
@rriveramcrus  I think these are great ideas! Once I have coverage for the API as it stands, I'll add these with stubbed tests, let me know if you're interested in collaborating on a driver/emulator that will leverage these functions.

rgundi commented on drivers/battery/sbs_gauge/sbs_gauge.h:
```diff
@@ -0,0 +1,80 @@
+#define SBS_GAUGE_CMD_REM_CAPACITY_ALARM    0x01 /* LowCapacityAlarmThreshold */
+#define SBS_GAUGE_CMD_REM_TIME_ALARM        0x02 /* RemainingTimeToEmptyThreshold */
+#define SBS_GAUGE_CMD_BATTERY_MODE          0x03 /* BatteryOperatingMode */
+#define SBS_GAUGE_CMD_AR                    0x04 /* AtRate */
```
Standard commands 0x0 to 0x4 are R/W. I don’t see a sbs_cmd_reg_write function in sbs_gauge.c.

aaronemassey replied:
This PR is focused on adding a fuel-gauge API while maintain feature parity with the sensor version of the sbs_gauge. But you're correct this was a mistake in the original sbs gauge driver. I'll add a todo here to add an sbs_gauge write.

aaronemassey replied:
This will also be part of an API extension in a future PR to allow some generic writing as well.

rgundi commented on drivers/battery/sbs_gauge/sbs_gauge.h:
```diff
@@ -0,0 +1,80 @@
+#define SBS_GAUGE_CMD_FC_TEMP               0x27 /* FCTemp */
+#define SBS_GAUGE_CMD_FC_STATUS             0x28 /* FCStatus */
+#define SBS_GAUGE_CMD_FC_MODE               0x29 /* FCMode */
+#define SBS_GAUGE_CMD_AUTO_SOFT_OFF         0x2A /* AutoSoftOff */
```
The Standard Commands 0x25 to 0x2A are reserved. As per the Revision 1.1 of the spec, “Reserved Command Error - 0x0002” needs to be returned if there is an attempt to read/write to these codes.

aaronemassey replied:
Good catch. I'll fix this.

rgundi replied:
Agree to this.

rgundi commented on drivers/battery/sbs_gauge/sbs_gauge.h:
```diff
@@ -0,0 +1,80 @@
+#define SBS_GAUGE_CMD_DEVICE_NAME           0x21 /* DeviceName */
+#define SBS_GAUGE_CMD_DEVICE_CHEM           0x22 /* DeviceChemistry */
+#define SBS_GAUGE_CMD_MANUFACTURER_DATA     0x23 /* ManufacturerData */
+#define SBS_GAUGE_CMD_DESIGN_MAX_POWER      0x24 /* DesignMaxPower */
```
Standard Command 0x24 is not defined at all in the Revision 1.1 of the spec.

aaronemassey replied:
I'll fix this too. Many of these errors were propagated from the origin sbs_gauge driver. I'll make a 2nd PR fixing that one. Thanks for pointing these out!

rgundi commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,200 @@
+enum battery_disconnect_state {
+	BATTERY_DISCONNECTED = 0,
+	BATTERY_NOT_DISCONNECTED,
+	BATTERY_DISCONNECT_ERROR,
```
Better to name it "battery_connect_state" and have only 2 values "connected" and "disconnected". Since this is mostly a GPIO, not sure what "battery_disconnect_error" can be.

rgundi commented on drivers/battery/sbs_gauge/sbs_gauge.h:
```diff
@@ -0,0 +1,80 @@
+#define SBS_GAUGE_CMD_CELL_V4               0x3C /* CellVoltage4 */
+#define SBS_GAUGE_CMD_CELL_V3               0x3D /* CellVoltage3 */
+#define SBS_GAUGE_CMD_CELL_V2               0x3E /* CellVoltage2 */
+#define SBS_GAUGE_CMD_CELL_V1               0x3F /* CellVoltage1 */
```
The standard commands 0x00, 0x2F, 0x3C, 0x3D, 0x3E and 0x3F are manufacture-defined functions. Different batteries may use different commands for the same functionality. A way to assign commands to these functionalities needs to be provided.

aaronemassey replied:
Adding in future PR.

rgundi commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,200 @@
+	/** Temperature in 0.1 K */
+	BATTERY_TEMPERATURE,
+	/** Battery voltage (mV) */
+	BATTERY_VOLTAGE,
```
Suggest battery_property_type to have all the commands mentioned in the smart battery spec. That way a separate API would not be needed for time_at_rate, time_to_empty and time_to_full.

rgundi commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,200 @@
+
+	battery_wait_for_stable wait_for_stable;
+	battery_sleep_fuel_gauge sleep_fuel_gauge;
+	battery_cut_off cut_off_charge;
```
Please include APIs to support below features. These implementations are manufacture specific.
a.	Enable battery shipment mode
b.	Enable EMShut mode

rgundi commented on drivers/battery/sbs_gauge/sbs_gauge.h:
```diff
@@ -0,0 +1,80 @@
+#define SBS_GAUGE_CMD_CELL_V3               0x3D /* CellVoltage3 */
+#define SBS_GAUGE_CMD_CELL_V2               0x3E /* CellVoltage2 */
+#define SBS_GAUGE_CMD_CELL_V1               0x3F /* CellVoltage1 */
+
```
There are some commands which are manufacture-defined and are not defined in the SBS (like Peak power, peak current etc). Different batteries may have different commands for the same functionality. A way to assign commands to these functionalities needs to be provided.

rgundi reviewed (commented):
Better to split the PR into multiple commits for easier understanding. For e.g.
a.	API definition
b.	Sample driver (sbs_gauge)
c.	Tests/examples of usage

aaronemassey committed:
test: SBS gauge gets initially supported props

Add a test that validates fetching every property from the sbs gauge driver
results in no driver error codes being returned.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

pacodinev commented:
An unwanted opinion from an inexperienced nobody.
This was also the opinion of another fellow user fabiobaltier in the previous PR.

In my opinion this API should take more inspiration from the Linux's [power_supply](https://docs.kernel.org/power/power_supply_class.html) subsystem. Also power_supply is not only used for implementing batteries, but also charges, adapters, etc. The Linux's API is mature, a lot of use cases and chargers/battery/coulomb counter ICs are implemented, even SBS. It is similar to Zephyr OS sensors API. It may need some rework, like event notifications, but I think it is still good enough.

[power_supply.h](https://elixir.bootlin.com/linux/v5.19.15/source/include/linux/power_supply.h)

teburd commented:
> An unwanted opinion from an inexperienced nobody. This was also the opinion of another fellow user fabiobaltier in the previous PR.
> 
> In my opinion this API should take more inspiration from the Linux's [power_supply](https://docs.kernel.org/power/power_supply_class.html) subsystem. Also power_supply is not only used for implementing batteries, but also charges, adapters, etc. The Linux's API is mature, a lot of use cases and chargers/battery/coulomb counter ICs are implemented, even SBS. It is similar to Zephyr OS sensors API. It may need some rework, like event notifications, but I think it is still good enough.
> 
> [power_supply.h](https://elixir.bootlin.com/linux/v5.19.15/source/include/linux/power_supply.h)

This really does seem like a great match, and honestly isn't that far away from what is being proposed here from all appearances.

It's kind of a neat idea though, to make this more than a battery centric API but a power supply API (where am I getting power from, how much current, whats the voltage, whats the source, battery? outlet? solar? if its a battery, which battery?).

Thanks for pointing this out!

teburd commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,152 @@
+	struct {
+		bool is_valid : 1;
+		bool use_cached : 1;
+		uint32_t property_type : 31;
```
```suggestion
		uint32_t property_type : 30;
```

teburd commented on drivers/battery/sbs_gauge/sbs_gauge.c:
```diff
@@ -0,0 +1,94 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <asm-generic/errno-base.h>
```
clangd auto include?

teburd commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,210 @@
+typedef int (*battery_get_property_t)(const struct device *dev, struct battery_get_property *props,
+				      size_t props_len);
+
+/* TODO(aaronmassey) Probably remove */
```
+1 to that, -ENOTSUP from get_property is fine for indicating this and makes the API even simpler

fabiobaltieri commented on drivers/battery/sbs_gauge/sbs_gauge.c:
```diff
@@ -0,0 +1,141 @@
+	return 0;
+}
+
+static int _sbs_gauge_get_prop(const struct device *dev, struct fuel_gauge_get_property *prop)
```
Drop the leading `_`.

fabiobaltieri commented on drivers/battery/sbs_gauge/sbs_gauge.h:
```diff
@@ -0,0 +1,80 @@
+#include <stdint.h>
+#include <zephyr/drivers/i2c.h>
+
+/*** Standard Commands ***/
```
More of curiosity... what does /*** do?

aaronemassey replied:
Copied from https://github.com/zephyrproject-rtos/zephyr/blob/33e583b6437e460006e20d04ef822b0954edb586/drivers/sensor/sbs_gauge/sbs_gauge.h#L13

You're right, we don't need the extra `*`.

fabiobaltieri commented on dts/bindings/battery/battery.yaml:
```diff
@@ -0,0 +1,84 @@
+description: |
+   Battery parameters and info
+
+   Voltage is in millivolts
+   Current is in milliamperes
+   Temperature is in Celsius degrees
```
Does this need to be specified for every driver? It's part of the API, I think it makes sense to have unit markers in the enum comment and probably even in the properties below, then this should be redundant.

aaronemassey replied:
So looking closer at this, I think the battery.yaml ought to be clearly defined in a later PR. I'm keeping this PR to defining a new fuel-gauge API with a sample driver and basic tests (which I'm happy to move to another PR, if other reviewers are happy with that).

fabiobaltieri commented on dts/bindings/battery/battery.yaml:
```diff
@@ -0,0 +1,84 @@
+      type: int
+   voltage_min:
+      description: |
+         Minimum voltage of the battery.
```
These description I think would be a better place to put units (rather than the comment for the whole binding)

fabiobaltieri commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,136 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifdef CONFIG_BATTERY
```
Uh... why? Don't think any other subsystem does this, can't think of a good reason to silently blank the file.

aaronemassey replied:
Good catch, pretty sure that was for early testing.

fabiobaltieri commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,136 @@
+	FUEL_GAUGE_VOLTAGE,
+
+	/** Reserved to demark downstream custom properties */
+	FUEL_GAUGE_CUSTOM_BEGIN = 256,
```
I think that if you do this the enum will always be 2 bytes. Anyway, how about doing it like the sensor API instead?

https://github.com/zephyrproject-rtos/zephyr/blob/33e583b6437e460006e20d04ef822b0954edb586/include/zephyr/drivers/sensor.h#L188-L197

That way there's only one gap.

fabiobaltieri commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,136 @@
+
+	/** Property field for getting */
+	union {
+		/** Fields have the format: */
+		/** FUEL_GAUGE_PROPERTY_FIELD */
+		/** type property_field; */
```
Is this meant to be a non-doxygen comment?

fabiobaltieri commented on include/zephyr/drivers/battery.h:
```diff
@@ -0,0 +1,136 @@
+		/** FUEL_GAUGE_PROPERTY_FIELD */
+		/** type property_field; */
+
+		/** Dynamic Battery Info */
```
Same.

fabiobaltieri reviewed (commented):
Hey some of the intermediate patches (the ones in `subsys/emul/i2c/emul_sbs_gauge.c`) are independent from the main one implementing the new subsystem right? I'd suggest moving them on a separate PR and getting them merged separately, keep this focused on the subsystem.

aaronemassey commented on dts/bindings/battery/battery.yaml:
```diff
@@ -0,0 +1,84 @@
+   Current is in milliamperes
+   Temperature is in Celsius degrees
+
+compatible: "zephyr,battery"
```
Changed `zephyr,battery` to `zephyr,fuel-gauge` to be more specific in language. Currently it's just a stub that's included by the experimental sbs-gauge compatible. I think that's okay for now.

aaronemassey commented:
@fabiobaltieri @pacodinev @teburd The direction I'd like to go with this work is to keep this PR small and focused to fuel-gauge ICs then we build on top of this a power supply API that's in userspace. Making an RFC with an image showing its long-term development. Part of this should also include a device driver API for charger-IC.

aaronemassey commented:
Blocked by https://github.com/zephyrproject-rtos/zephyr/pull/52180

aaronemassey committed:
fuel_gauge: Sample sbs gauge driver with tests

Add a sample sbs gauge driver with feature parity and basic tests
comparison to its sensor counter-part. Includes a simple stub test that is
extended upon.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey commented:
Removed the test for failing registers as I'll add that in its own tiny PR as it introduces some test mockable.

aaronemassey commented:
RFC Updated https://github.com/zephyrproject-rtos/zephyr/issues/44847

aaronemassey commented:
Tried adding this API as "experimental" but looks like that's done at release time?

teburd commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -0,0 +1,135 @@
+		/** FUEL_GAUGE_FULL_CHARGE_CAPACITY */
+		uint32_t full_charge_capacity;
+		/** FUEL_GAUGE_IMBALANCE_MV */
+		uint32_t imbalance_mv;
```
I didn't see this property listed above

teburd commented on tests/drivers/fuel_gauge/sbs_gauge/src/test_sbs_gauge.c:
```diff
@@ -0,0 +1,89 @@
+{
+	/* Validate what props are supported by the driver */
+
+	struct fuel_gauge_get_property props[] = {
```
IMBALANCE isn't used here either

teburd commented:
Minor things I caught here, but this looks great otherwise. The units are probably different than SBS, but Zephyr is used in a broad spectrum of devices and 32bit values in micro voltage/amps still can represent kilovolts and kiloamps which is far higher than a fuel gauge is likely to ever support, unless you are talking about fast car DC chargers... and even then, kilo's is enough.

teburd commented:
And of course in reviewing there's a new sensor PR for a new gauge :-) #52349

teburd commented:
> Tried adding this API as "experimental" but looks like that's done at release time?

Add it to the API docs page and the status of it in the PR with the next release in mind

yperess commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -0,0 +1,135 @@
+	 * See driver instance header for details on implementation and
+	 * how the average is calculated. Units in mA negative=discharging
+	 */
+	FUEL_GAUGE_AVG_CURRENT = 0,
+
+	/** Battery current (mA); negative=discharging */
+	FUEL_GAUGE_CURRENT,
+	/** Whether the battery underlying the fuel-gauge is cut off from charge */
+	FUEL_GAUGE_CHARGE_CUTOFF,
+	/** Cycle count in 1/100ths (number of charge/discharge cycles) */
+	FUEL_GAUGE_CYCLE_COUNT,
+	/** Connect state of battery */
+	FUEL_GAUGE_CONNECT_STATE,
+	/** General Error/Runtime Flags */
+	FUEL_GAUGE_FLAGS,
+	/** Full Charge Capacity in mAh (might change in some implementations to determine wear) */
+	FUEL_GAUGE_FULL_CHARGE_CAPACITY,
+	/** Is the battery physically present */
+	FUEL_GAUGE_PRESENT_STATE,
+	/** Remaining capacity in mAh */
+	FUEL_GAUGE_REMAINING_CAPACITY,
+	/** Remaining battery life time in minutes */
+	FUEL_GAUGE_RUNTIME_TO_EMPTY,
+	/** Remaining time in minutes until battery reaches full charge */
+	FUEL_GAUGE_RUNTIME_TO_FULL,
+	/** Absolute state of charge (percent, 0-100) - expressed as % of design capacity */
+	FUEL_GAUGE_STATE_OF_CHARGE,
+	/** Temperature in 0.1 K */
+	FUEL_GAUGE_TEMPERATURE,
+	/** Battery voltage (mV) */
+	FUEL_GAUGE_VOLTAGE,
+
+	/** Reserved to demark end of common fuel gauge properties*/
+	FUEL_GAUGE_COMMON_COUNT,
+	/** Reserved to demark downstream custom properties */
+	FUEL_GAUGE_CUSTOM_BEGIN = FUEL_GAUGE_COMMON_COUNT,
+
+	/** Reserved to demark end of valid enum properties */
+	FUEL_GAUGE_PROP_MAX = UINT16_MAX
+};
```
Verified, if you use C++ downstream users will not be able to do something like:

```
#define MY_PROPERTY FUEL_GAUGE_CUSTOM_BEGIN
```

And pass it as the enum, C++ is more strongly typed. It is possible to include the cast in the #define to look like:
```
#define MY_PROPERTY ((enum fuel_gauge_property_type)FUEL_GAUGE_CUSTOM_BEGIN)
```

But I think that at this point you're losing the type safety anyway. We should just use #define for all these values instead.

yperess replied:
Actually I played with this some more and I think we're far more likely to add properties here. Let's keep the enum

aaronemassey replied:
Thanks for testing and pointing this out. I've converted the enum to `#define`s.

aaronemassey commented:
> Minor things I caught here, but this looks great otherwise. The units are probably different than SBS, but Zephyr is used in a broad spectrum of devices and 32bit values in micro voltage/amps still can represent kilovolts and kiloamps which is far higher than a fuel gauge is likely to ever support, unless you are talking about fast car DC chargers and batteries... and even then, kilo's is probably enough.

Thanks!Moved to uA/uV.

aaronemassey committed:
fuel_gauge: Add basic stub API doc.

Add a short stub doc as a placeholder for future documentation on the fuel
gauge API.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

teburd commented on dts/bindings/battery/fuel-gauge.yaml:
```diff
@@ -0,0 +1,12 @@
+#
```
File path for this still including battery, if the goal is fuel-gauge and charger under a power supply umbrella, then the path should match probably

aaronemassey replied:
Whoops.

teburd reviewed (changes requested):
Last thing I missed... the dts binding path is probably not quite right.

Also add this API to the doc/develop/api/overview.rst as experimental as of the next release.

teburd commented:
Don't forget "Also add this API to the doc/develop/api/overview.rst as experimental as of the next release."

aaronemassey commented:
@teburd 
> Don't forget "Also add this API to the doc/develop/api/overview.rst as experimental as of the next release."

Where's a good location to add this? In `fuel_gauge.rst`?

teburd commented:
> @teburd
> 
> > Don't forget "Also add this API to the doc/develop/api/overview.rst as experimental as of the next release."
> 
> Where's a good location to add this? In `fuel_gauge.rst`?

Added to the table in the file
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/api/overview.rst

aaronemassey commented:
> > @teburd
> > > Don't forget "Also add this API to the doc/develop/api/overview.rst as experimental as of the next release."
> > 
> > 
> > Where's a good location to add this? In `fuel_gauge.rst`?
> 
> Added to the table in the file https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/api/overview.rst

Oh. Somehow I opened this file up in read-only mode so I thought I wasn't to make changes against it. Thanks!

pacodinev commented:
Disclaimer: I am very inexperienced, so you should probably not waste your time with me! :)
This is based on my opinion and I am sorry for being this long.
From what I am seeing, I might be late to the party. :(

Last time, I didn't saw the RFC, but now I understand it.

My biggest issue with that RFC (and that PR) is the rarity of separate charger IC and fuel gauge IC in the embedded world. Most boards that feature some kind of battery related IC have either fuel gauge IC, charger IC or an IC that combines them both. Example is [MAX77818](https://www.maximintegrated.com/en/products/power/battery-management/MAX77818.html).
This device has to have two separate drivers, because it is both a fuel gauge and a charger. This isn't impossible (Linux's MFD), but I think it is unneeded complication, at least compared to Linux's power_supply.

Yes, I know, SBS is an example of separate charger and fuel gauge (SBS battery), but firstly rarely embedded device uses SBS, and secondly if you want to change charging parameters (at least in the case you can), (most of the time) you need to communicate the change to the SBS battery, and then the SBS battery will communicate the changes to the charger.

As a last example, the very integrated [AXP202](https://dl.linux-sunxi.org/AXP/AXP202%20Datasheet_v1.0_en.pdf) from X-Powers.
Linux gets away with using 3 instances of power_supply driver: battery, ac-power, usb-power. (I am not counting the regulator interface and etc.)
With this RFC, you will need to use 4 instances of 3 drivers: 1 - battery, 1 - charger and 2 of something else for usb and ac power monitoring, for the same functionality. And they all will have some very similar APIs.
And in both cases, you will need to have some userspace driver for more advanced processing.
